li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/producer/LiKafkaFederatedProducerImpl.java

```diff
@@ -0,0 +1,274 @@
+/*
+ * Copyright 2017 LinkedIn Corp. Licensed under the BSD 2-Clause License (the "License").  See License in the project root for license information.
+ */
+
+package com.linkedin.kafka.clients.producer;
+
+import com.linkedin.kafka.clients.auditing.Auditor;
+import com.linkedin.kafka.clients.common.ClusterGroupDescriptor;
+import com.linkedin.kafka.clients.largemessage.LargeMessageSegment;
+import com.linkedin.kafka.clients.largemessage.MessageSplitter;
+import com.linkedin.kafka.clients.largemessage.MessageSplitterImpl;
+import com.linkedin.kafka.clients.metadataservice.MetadataServiceClient;
+
+import java.io.IOException;
+import java.lang.reflect.Method;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Properties;
+import java.util.Set;
+import java.util.UUID;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import org.apache.kafka.clients.producer.Callback;
+import org.apache.kafka.clients.producer.KafkaProducer;
+import org.apache.kafka.clients.producer.Producer;
+import org.apache.kafka.clients.producer.ProducerConfig;
+import org.apache.kafka.clients.producer.ProducerRecord;
+import org.apache.kafka.clients.producer.RecordMetadata;
+import org.apache.kafka.clients.consumer.OffsetAndMetadata;
+import org.apache.kafka.common.Metric;
+import org.apache.kafka.common.MetricName;
+import org.apache.kafka.common.PartitionInfo;
+import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.common.errors.ProducerFencedException;
+import org.apache.kafka.common.serialization.ByteArraySerializer;
+import org.apache.kafka.common.serialization.Serializer;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * This is a producer implementation that works with a federated Kafka cluster, which consists of one or more physical
+ * Kafka clusters.
+ */
+public class LiKafkaFederatedProducerImpl<K, V> implements LiKafkaProducer<K, V> {
+  private static final Logger LOG = LoggerFactory.getLogger(LiKafkaFederatedProducerImpl.class);
+
+  // Per-cluster producer context
+  private class PerClusterProducerContext<K, V> {
+    private Producer<byte[], byte[]> _producer;
+    /*package private for testing*/ Auditor<K, V> _auditor;
+
+    // A counter of the threads in the middle of sending messages. This is needed to ensure when we close the producer
+    // everything is audited.
+    private final AtomicInteger _numThreadsInSend = new AtomicInteger();
```

Comment:
performance nit - LongAdder would be cheaper for this

Reply:
Done